docs: new tutorial "generate code" (#1025)
asyncapi/website: pages/docs/tutorials/generating-code.md

---
title: Streetlights - Generating Code 
description: In this tutorial, you'll learn how to generate your code from an AsyncAPI document.
weight: 130
---

# Introduction

In this tutorial, you'll learn how to generate your code from an AsyncAPI document using the AsyncAPI generator

# Background Context

> The [AsyncAPI Generator](https://github.com/asyncapi/generator) tool is used to create the documentation from the specification file. It is built in JavaScript and can be accessible using npm or Docker. 

Reviewer: ```suggestion
The [AsyncAPI Generator](https://github.com/asyncapi/generator) tool that you can use to generate whatever you want based on the AsyncAPI document. You can generate docs and code. It can be used as a library in Nodejs application or through the [AsyncAPI CLI](https://github.com/asyncapi/cli).
```

> The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates helps to specify what exactly must be generated and for this tutorial, a Nodejs template will be specified.

Reviewer: ```suggestion
The generator tool supports a number of templates to generate code for a variety of different languages and protocols as the output. These templates help to specify what exactly must be generated, and in this tutorial, you use [Nodejs template](https://github.com/asyncapi/nodejs-template).
```

there are different schools for writing docs (tutorials, to be specific). I prefer school:
- where future tenses are not used, so you basically describe what "is happening in the tutorial" and not "what will happen"
- I prefer when tutorials do not say what `we do` but what `you do`, so they are directed towards user and engage the user. I prefer `you` but I know there are people that prefer `we` as they say that is like telling users they are not on their own 😄 

I'm just justifying my proposal for a change. My preference doesn't matter much, more important is that we should stick in all tutorials to one and be consistent.

@alequetzalli I leave it up to you folks but feels like it makes sense to create a kind of check list of rules, that can also be used in the review

Reviewer 2: 1. Both `you` and `we` are great to use in a tutorial because they accomplish different useful goals (which Lukasz already mentioned.) That said, I agree that consistency is key and it's always nice to choose one and stick to that selection. I think using `you` is great in our tutorials, but I don't want it to be "wrong" to sometimes use the word "we". 😂 

2. Yes, tutorial steps should be actions taken in **present** tense.

> # Installation Guide

Reviewer: I see the guide shows generator through the CLI. We should focus on using https://github.com/asyncapi/cli and stick to just one way of installing depending on your operating system. So if you use Windows, then probably `npm` will stay, but if you use `Macos` then promote here `Macos` way of doing things, and then just refer that there are other options to install, and link to https://github.com/asyncapi/cli#installation

what confuses me that this section is focused on installation, but the, you install CLI in another section, in code generation

Before proceeding to the next step, make sure the following are still actively installed on your local PC:

1. Install [Node.js](https://nodejs.org/en/download/) (v15 or newer).

2. Install Git on your Operating System:
    - [Install Git on MacOs](https://git-scm.com/download/mac)
    - [Install Git on Windows](https://git-scm.com/download/win)
    - [Install Git on Linux](https://git-scm.com/download/linux)

> # Generating code

Reviewer: max `##` should be used in docs, as it will clash with `#` (h1) on the website

> # Generating code

Reviewer: maybe Generate publisher application (as it will related to the other section I suggested below)

In this section, we will go ahead with the steps to enable us generate our code from the document;

> ### 1. Install the generator to use it as a command-line(CLI) tool

Reviewer: I see you follow this approach from original tutorial, 

I did it like it, and I'm super unhappy with it, I recommend not doing headings for steps
look how ugly it is here https://www.asyncapi.com/docs/tutorials/streetlights, especially the TOC on the rights is unusable

feel free to change that

<CodeBlock language="bash">
{`npm install -g @asyncapi/generator`}
</CodeBlock>

The command above helps install the generator tool globally.

> ### 2. Create a directory for your projects and enter it:

Reviewer: as far as I'm aware, the idea is that this is one of 4 tutorials that is created as a replacement for existing streetlight.

so before this tutorial, there will be "creation of AsyncAPI doc" and "validation of AsyncAPI doc"

so we should not create AsyncAPI file here again, not paste it here again.

imho we should reference previous tutorials 

we can only add a note that if someone bypassed previous tutorials, they can quickly generate AsyncAPI file for them using proper AsyncAPI CLI command

<CodeBlock language="yaml">
{`cat <<EOT >> asyncapi.yaml
asyncapi: '2.5.0'
info:
  title: Streetlights API
  version: '1.0.0'
  description: |
    The Smartylighting Streetlights API allows you
    to remotely manage the city lights.
  license:
    name: Apache 2.0
    url: 'https://www.apache.org/licenses/LICENSE-2.0'
servers:
  mosquitto:
    url: mqtt://test.mosquitto.org
    protocol: mqtt
channels:
  light/measured:
    publish:
      summary: Inform about environmental lighting conditions for a particular streetlight.
      operationId: onLightMeasured
      message:
        name: LightMeasured
        payload:
          type: object
          properties:
            id:
              type: integer
              minimum: 0
              description: Id of the streetlight.
            lumens:
              type: integer
              minimum: 0
              description: Light intensity measured in lumens.
            sentAt:
              type: string
              format: date-time
              description: Date and time when the message was sent.
EOT`}
</CodeBlock>

### 4. Trigger generation of the Node.js code:
<CodeBlock language="bash">

> {`ag asyncapi.yaml @asyncapi/nodejs-template -o output -p server=mosquitto`}

Reviewer: please use new AsyncAPI CLI command - examples here https://github.com/asyncapi/generator/pull/842/

</CodeBlock>

From the command above, the `-o` specifies where to output the result.

### 5. And voilà! List all files in directory and check that Node.js application is generated:
<CodeBlock language="bash">
{`cd output && ls`}
</CodeBlock>

# Running your code

### 1. Install dependencies of newly generated application:
<CodeBlock language="bash">
{`npm install`}
</CodeBlock>

### 2. Start the application:
<CodeBlock language="bash">
{`npm start`}
</CodeBlock>

> ### 3. In another terminal install the MQTT.js library:

Reviewer: better add new section like `Send a message to subscriber`

and you will explain that to produce a message, users will use `MQTT client` that needs to be installed first

<CodeBlock language="bash">
{`npm install mqtt -g`}
</CodeBlock>

> ### 4. Send a correct message to your application:

Reviewer: about 4 and 5, these are about sending good and bad message, just focus on one, not calling it `correct`, because the 4th tutorial is about message validation
